src/test/java/org/inferred/freebuilder/processor/util/testing/BehaviorTestRunner.java

```diff
@@ -0,0 +1,475 @@
+    if (tester != null) {
+      // Only one child, no pre-run needed.
+      super.runChild(method, notifier);
+      return;
+    }
+    Child child = new Child(method);
+    tester = child;
+    try {
+      super.runChild(method, new RunNotifier() {
+        @Override
+        public void fireTestFailure(Failure failure) {
+          if (COMPILATION_FAILURE_EXPECTED.equals(failure.getMessage())) {
+            // Test is expecting compilation to fail, so merging with other tests will only
+            // propagate the failure. Mark the test as unmergeable.
+            child.unmergeable = true;
+          }
+        }
+      });
+    } finally {
+      tester = null;
+      children.add(child);
+    }
+  }
+
+  /**
+   * Runs {@code child} with {@code sharedCompiler}.
+   */
+  private void runChild(
+      RunNotifier notifier,
+      SharedCompiler sharedCompiler,
+      FrameworkMethod child) {
+    tester = new DelegatingBehaviorTester(sharedCompiler);
+    try {
+      super.runChild(child, notifier);
+    } finally {
+      tester = null;
+    }
+  }
+
+  /**
+   * Returns a new fixture for running a test. Executes the test class's no-argument constructor,
+   * then injects the correct BehaviorTester implementation.
+   */
+  @Override
+  public Object createTest() throws Exception {
+    checkState(tester != null);
+    Object test = super.createTest();
+    // Inject the correct BehaviorTester implementation.
+    testerField.set(test, tester);
+    return test;
+  }
+
+  /**
+   * Groups children that can share a compiler.
+   */
+  private static Queue<SharedCompiler> shareCompilers(List<Child> children) {
+    Queue<SharedCompiler> sharedCompilers = new ArrayDeque<>();
+    for (Child child : children) {
+      Optional<SharedCompiler> sharedCompiler = sharedCompilers
+          .stream()
+          .filter(c -> c.canShareCompiler(child))
+          .findAny();
+      if (sharedCompiler.isPresent()) {
+        sharedCompiler.get().addChild(child);
+      } else {
+        sharedCompilers.add(new SharedCompiler(child));
+      }
+    }
+    return sharedCompilers;
+  }
+
+  /**
+   * Fails the {@link #introspection} test unless we managed to share a compiler.
+   */
+  private void verifyCompilerShared(RunNotifier notifier, int numChildren, int numCompilations) {
+    System.out.println(String.format(
+        "Merged %d tests into %d compiler passes",
+        numChildren,
+        numCompilations));
+    int numFilteredChildren = super.getDescription().getChildren().size();
+    if (numFilteredChildren == numChildren) {
+      if (numChildren == numCompilations) {
+        notifier.fireTestFailure(new Failure(
+            introspection, new AssertionError(MERGE_FAILURE_MESSAGE)));
+      }
+    }
+  }
+
+  /**
+   * A BehaviorTester that just remembers the processors, units and tests it was asked to compile.
+   */
+  private static class Child implements BehaviorTester {
```

Comment:
this class is a bit hefty. Any chance you could factor out these subclasses? Would allow them to be tested separately as well.

Feel it's a *little* bit bad to not test test code, but don't feel strongly.